File: 02.py
#!/usr/bin/env python3

from enum import Enum

class RPS(Enum):
    ROCK = 1
    PAPER = 2
    SCISSORS = 3

order = [RPS.ROCK, RPS.PAPER, RPS.SCISSORS]

def charToRPS(char):
    if (char == "A" or char == "X"):
        return RPS.ROCK
    if (char == "B" or char == "Y"):
        return RPS.PAPER
    if (char == "C" or char == "Z"):
        return RPS.SCISSORS

def getLosingMove(enemyMove):
    return order[(order.index(enemyMove) - 1) % 3]

def getWinningMove(enemyMove):
    return order[(order.index(enemyMove) + 1) % 3]

def score(enemy, myThrow):
    total = 0
    total += myThrow.value
    if (getWinningMove(enemy) == myThrow):
        total += 6
    if (enemy == myThrow):
        total += 3
    return total

with open("./02-input.txt", "r") as file:
    total = 0
    for line in file:
        left, right = line.strip().split(" ")
        instscore = score(charToRPS(left), charToRPS(right))
        print(instscore)
        total += instscore

    print(total)

File: 02.clj
(ns adventofcode2022.02
  (:require [clojure.string :as s]))


(def moves [:r :p :s])
(.indexOf moves :r)

(defn getWinningMove
  [enemyMove]
  (nth (cycle moves) (+ 4 (.indexOf moves enemyMove))))

(defn getLosingMove
  [enemyMove]
  (nth (cycle moves) (+ 2 (.indexOf moves enemyMove))))

(defn str->RPS
  [str]
  (case str
      ("A" "X") :r
      ("B" "Y") :p
      ("C" "Z") :s))

(defn score
  [[enemy mine]]
  (+ (cond
       (= mine (getWinningMove enemy))   6
       (= enemy mine) 3
       :else 0)

     (inc (.indexOf moves mine))))

(defn part1
  []
  (->>
   (slurp "02-input.txt")
   (s/split-lines)
   (map s/trim)
   (map #(s/split % #" "))
   (map #(map str->RPS %))
   (map score)
   (apply +)))

(defn getMove
  [enemyMove instruction]
  (case instruction
    "X" (getLosingMove enemyMove)
    "Y" enemyMove
    "Z" (getWinningMove enemyMove)))

(defn strTuple->RPSInstruction
  [[enemyStr instruction]]
  (let [enemyMove (str->RPS enemyStr)]
    [enemyMove (getMove enemyMove instruction)]))

(defn part2
  []
  (->>
   (slurp "02-input.txt")
   (s/split-lines)
   (map s/trim)
   (map #(s/split % #" "))
   (map strTuple->RPSInstruction)
   (map score)
   (apply +)))

(part1)
(part2)
